state/Wait: Fix race when reading exit status

vvoland opened the pull request:
Fixes:
- https://github.com/moby/moby/issues/39352

Before this change there was a race condition between State.Wait reading
the exit code from State and the State being changed immediately after the
change which ended the State.Wait.

Now, each State.Wait call holds a pointer to the StateStatus obtained
from State at the time of State.Wait call. This makes it possible to
read the correct exit code, even if the State was already modified.

The issue caused the `docker run --restart=always ...' to sometimes exit
with 0 exit code, because the process was already restarted by the time
State.Wait got the chance to read the exit code.

Test run
Before:
```
$ go test -count 1 -run TestCorrectStateWaitResultAfterRestart .
--- FAIL: TestCorrectStateWaitResultAfterRestart (0.00s)
    state_test.go:198: expected exit code 10, got 0
FAIL
FAIL    github.com/docker/docker/container      0.011s
FAIL

```

After:
```
$ go test -count 1 -run TestCorrectStateWaitResultAfterRestart .
ok      github.com/docker/docker/container      0.011s
```

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

**- What I did**
Fixed State.Wait returning unexpected StateStatus

**- How I did it**
Store StateStatus for Wait and make each Wait's goroutine store a pointer to it. The subsequent State modifications will allocate a new StateStatus, so the old one won't be affected. 

**- How to verify it**
Run tests:
```
cd container
go test -count 1 -run TestCorrectStateWaitResultAfterRestart .
```
or

(it's faster if you run multiple of those with different names)
```
while [ true ]; do
  docker run --name test --restart=always busybox sh -c 'exit 11'
  if [ ! $? -eq 11 ]; then
    echo "Reproduced!"
    break
   fi
   docker rm -f test
done
```
**- Description for the changelog**
Fix `docker run` with restart policy to sometimes exit with 0 code

**- A picture of a cute animal (not mandatory but encouraged)**
![gimmebananahuman](https://user-images.githubusercontent.com/5046555/176409011-6cdc8460-6c4b-4008-802e-a35c406f8991.jpg)

thaJeztah commented:
Interesting failure; wondering if that test was actually trying to test something that shouldn't be supported, but happened to work because of this;

```
=== RUN   TestDockerSuite/TestRestartAutoRemoveContainer
    docker_cli_restart_test.go:330: assertion failed: 
        Command:  /usr/local/cli/docker restart ce5a4d9591f573ce91aafda5dd28761cb808eaac0f4465d50e7736b2f7926bc0
        ExitCode: 1
        Error:    exit status 1
        Stdout:   
        Stderr:   Error response from daemon: Cannot restart container ce5a4d9591f573ce91aafda5dd28761cb808eaac0f4465d50e7736b2f7926bc0: container is marked for removal and cannot be started
        
        
        Failures:
        ExitCode was 1 expected 0
        Expected no error
    --- FAIL: TestDockerSuite/TestRestartAutoRemoveContainer (0.52s)
```

Will have to dig up history for that test to see what it was added for

thaJeztah commented:
test was added in https://github.com/moby/moby/commit/3c2886d8a45d8e79b00ab413d91f1af871b58d0a (https://github.com/moby/moby/pull/20848), and originally looked like;

```go
func (s *DockerSuite) TestRestartAutoRmoveContainer(c *check.C) {
 	out, _ := runSleepingContainer(c, "--rm")

 	id := strings.TrimSpace(string(out))
 	dockerCmd(c, "restart", id)
 	err := waitInspect(id, "{{ .State.Restarting }} {{ .State.Running }}", "false true", 15*time.Second)
 	c.Assert(err, checker.IsNil)

 	out, _ = dockerCmd(c, "ps")
 	c.Assert(out, checker.Contains, id[:12], check.Commentf("container should be restarted instead of removed: %v", out))
 }
```

thaJeztah commented:
So, I _think_ the flow for that situation was that on `docker restart`, we mark the container as restarting, and/or disable monitoring of the container's state to prevent it being deleted in that `stop/start` cycle.

vvoland commented:
There was another race, when the AutoRemove could be restored to true before the auto removal was actually considered by the monitor.
I also added an integration test `TestDaemonRestartAutoRemove` that should check for the same as `TestDockerSuite/TestRestartAutoRemoveContainer` but it's easier to run locally than integration-cli (also our CI seems to run integration-cli tests only on Windows?).
Should I delete TestDockerSuite/TestRestartAutoRemoveContainer now or is it fine to leave it?

thaJeztah commented on daemon/restart.go:
```diff
@@ -54,12 +54,23 @@ func (daemon *Daemon) containerRestart(ctx context.Context, container *container
+		// Do it after the container has been started (or the stop failed), so it doesnt'
+		// get restored to true BEFORE the auto removal is actually considered by the monitor.
+		if restoreAutoRemove {
+			defer func() {
```
If we call this in a `defer`, it will be called after `daemon.checkpointAndSave`, which I think will mean that the container configuration (on disk) would now always have `AutoRemove: false` 🤔

thaJeztah replied:
That said; I wonder if we could use the `Restarting` state of the container somehow; there already seems to be some code for that, so it's a bit confusing why we need to also set the auto-remove option; https://github.com/moby/moby/blob/7b9275c0da707b030e62c96b679a976f31f929d3/daemon/monitor.go#L65-L71

thaJeztah replied:
The `err == nil` in the first branch of the above looks somewhat fishy; it would mean that if an error occurs, we'll go ahead and stop (and remove) the container; perhaps that's ok, but wondering what errors can occur.

There's also a `daemon.autoRemove()` further down in that same function (logic is confusing and complicated)

vvoland replied:
The err here is not container error, but error returned from `ShouldRestart`. err is not nil for example when restart is cancelled for some reason, for example the container with restart policy was stopped manually.
I included some more logging here and moved handling the err a bit closer to the ShouldRestart call.

vvoland committed:
integration: TestWaitRestartedContainer

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
integration: Add TestContainerWithAutoRemoveCanBeRestarted

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland commented:
AutoRemove is no longer mutated and so calling `checkpointAndSave` is also no longer needed. 

The start logic could be shared with the restart invoked by restart policy - this way `containerRestart` would set the `HasBeenManuallyRestarted` flag and call the `containerStop`, which would then be picked up by the monitor's `handleContainerExit`. We would need to implement some kind of WaitConditionRunning though - because we would need `containterRestart` to block until the container is started and without calling the directly `containerStart` we wouldn't know that.
I can post a follow-up draft tomorrow if that's something you think is worth working on.

thaJeztah commented:
Failure on Linux/amd64 (perhaps test needs updating?)

```
=== RUN   TestContainerWithAutoRemoveCanBeRestarted/stop
    restart_test.go:204: assertion failed: exited (inspect.State.Status string) != removing (string)
    --- FAIL: TestContainerWithAutoRemoveCanBeRestarted/stop (1.28s)
```

And on Windows; integer overflow?

```
=== FAIL: github.com/docker/docker/integration/container TestWaitRestartedContainer/next-exit (12.57s)
    wait_test.go:208: assertion failed: 5 (int64) != 3221226219 (waitRes.StatusCode int64)
    --- FAIL: TestWaitRestartedContainer/next-exit (12.57s)

=== FAIL: github.com/docker/docker/integration/container TestWaitRestartedContainer/default (12.73s)
    wait_test.go:208: assertion failed: 5 (int64) != 3221226219 (waitRes.StatusCode int64)
    --- FAIL: TestWaitRestartedContainer/default (12.73s)

=== FAIL: github.com/docker/docker/integration/container TestWaitRestartedContainer/not-running (9.82s)
    wait_test.go:208: assertion failed: 5 (int64) != 3221226219 (waitRes.StatusCode int64)
    --- FAIL: TestWaitRestartedContainer/not-running (9.82s)
```

vvoland commented:
The Windows case is caused by the fact that the `trap "exit 5" SIGTERM` doesn't work because the Windows based busybox container doesn't receive the SIGTERM, so the Stop timeouts and kills the container which exits with 3221226219 (0xC00002EB = STATUS_SYSTEM_SHUTDOWN).

Added a small test workaround for this Windows case because I couldn't find a way to capture the graceful exit signal.
But maybe there is some way? There probably should 😄

thaJeztah commented:
Quick search showed me https://github.com/containerd/containerd/issues/6834 / https://github.com/containerd/containerd/pull/6842 (admitted, that looks to be for the CRI API)

thaJeztah commented on container/container.go:
```diff
@@ -81,17 +81,18 @@ type Container struct {
+	RestartCount             int
+	HasBeenStartedBefore     bool
+	HasBeenManuallyStopped   bool // used for unless-stopped restart policy
+	HasBeenManuallyRestarted bool // used to distinguish restart caused by restart policy from the manual one
```
This `Container` struct is also used to store state on disk.

Do we need `HasBeenManuallyRestarted` to be persisted on disk? If not (i.e., if we only need it "in memory" for the duration of the restart), then we should exclude it from the JSON;

```suggestion
	HasBeenManuallyRestarted bool `json:"-"` // used to distinguish restart caused by restart policy from the manual one
```

vvoland replied:
Yeah I don't think we need it. Thanks!

thaJeztah commented on container/state.go:
```diff
@@ -209,6 +213,9 @@ func (s *State) Wait(ctx context.Context, condition WaitCondition) <-chan StateS
 
 	resultC := make(chan StateStatus, 1)
 
+	stopResult := s.waitStopResult
+	removeResult := s.waitRemoveResult
```
Perhaps silly question; are there cases where we handle _both_  `waitStopResult` and `waitRemoveResult` ? Or are they effectively mutually exclusive (stop -> get result, remove -> get result); if that's the case we could have a single `waitResult` that's used by both, or would that make the code more complicated?

vvoland replied:
The final result will be only one of them (whichever comes first). I think having only one result could also work. Will check.

thaJeztah commented on container/state.go:
```diff
@@ -292,6 +292,9 @@ func (s *State) SetStopped(exitStatus *ExitStatus) {
 	s.ExitCodeValue = exitStatus.ExitCode
 	s.OOMKilled = exitStatus.OOMKilled
+
+	s.saveWaitResult(s.waitStopResult)
+	s.waitStopResult = new(StateStatus)
```
I'm trying to get my head around these;

- we're passing a reference to a property of `s` to a method of `s`
- update the passed reference
- then immediately overwrite it with a new value

```go
func (s *State) saveWaitResult(result *StateStatus) {
 	result.exitCode = s.ExitCodeValue
 	result.err = s.Err()
 }
```


So effectively that's equivalent to;

```go
func (s *State) saveWaitResult() {
	s.waitStopResult = &StateStatus{
		exitCode: s.ExitCodeValue,
		err:      s.Err(),
	}
}
```

Which would require two separate functions for `saveWaitStopResult()` and `saveWaitRemoveResult()`, unless we can use the same (see my earlier comment)


Even then, I'm somewhat confused by the line following it;

```go
s.waitStopResult = new(StateStatus)
```

vvoland replied:
They are not equivalent.
`saveWaitResult` updates the value under reference, not the reference itself.
It's intentional - the State also holds the statuses as pointers. 
What `Wait` now does is stores a copy of these pointers which represent a StateStatus that is not yet filled, but will be filled once the corresponding wait channel is closed. 

```go
stopResult := s.waitStopResult
removeResult := s.waitRemoveResult
```

So `saveWaitResult` writes result to a StateStatus that `Wait` calls have pointer to. After that creates a fresh StateStatus that will be filled by the next exit/removal - this still doesn't affect waiting `Wait` calls, because they still have a pointer to the StateStatus that they are watching. 
This wouldn't be the case with your `saveWaitResult`, because it doesn't modify the StateStatus, it creates a new one.

I agree though that it's not really obvious in the code - maybe there is some cleaner way to express this?

vvoland commented:
Thanks for the review! The Windows error seems to go away without `-t` (docker run), so possibly we can keep these tests on Windows. Will try to kick the CI a few times to check if there aren't some other funny Windows errors.

thaJeztah commented:
Thanks! I see you updated the PR; looks like we have a new failure now (hope it wasn't because I made a wrong suggestion 😓)

```
=== FAIL: github.com/docker/docker/container TestStateRunStop (0.00s)
    state_test.go:120: Removal wait exitCode 1, expected 2, err %!q(<nil>)
```


Perhaps others are able to help review this one as well (@corhere ? 🤗); for easier review; look at the diff with whitespace changes hidden (add `?w=1`), which makes the diff easier to read; https://github.com/moby/moby/pull/43751/files?w=1

vvoland commented:
Yeah, sharing the result for both removal and stop breaks the case where you have multiple Waits for different scenarios and the result is overwritten by the other.

A small summary of the test failure:
```go
removalWait := s.Wait(ctx, WaitConditionRemoved)
exitWait := s.Wait(ctx, WaitConditionNextExit)

s.SetStopped(&ExitStatus{ExitCode: 1}) // Result is set for both waits, wakes the exitWait

s.SetRunning()

// Stop and remove container
s.SetStopped(&ExitStatus{ExitCode: 2}) // This is written to a fresh Result, so the removalWait won't get this
s.SetRemoved() // Wakes the removalWait, but it still has the ExitCode 1, instead of 2 which was associated with the removal
```

Reverting that one.

vvoland commented:
As for the event fanout with channels - actually my first attempt at fixing this looked similiar but then I just thought of this and decided to go with it because it seemed much cleaner. I will try the approach you suggested and we'll see how it looks. Thanks for the suggestion and great resource about Go concurrency!

Regarding making the `Wait`  synchronous - this involves much more code to be changed, so that's possibly something that could be addressed in a follow-up.
However I'm not sure if this wouldn't actually encourage a racy approach to calling it. Usually `Wait` involves some kind of action that will actually make the state transition. Let's take an restart for example:
 
```go
client.ContainerRestart(ctx, containerID)
c.WaitSynchronous(ctx, WaitConditionNextExit)
```

In this case, the container *could* already be restarted and in running state before `Wait` is even entered. 
So the correct approach here would involve adding the asynchronous behaviour explicitly:
```go
c := make(chan Result)
go func () {
  c <- ctr.WaitSynchronous(ctx, WaitConditionNextExit)
}()
client.ContainerRestart(ctx, containerID)
<- c
```

And the current asynchronous version makes it more natural to use the Wait safely: 
```go
c := ctr.Wait(ctx, WaitConditionNextExit)
client.ContainerRestart(ctx, containerID)
<- c
```

vvoland commented:
> > Yeah, sharing the result for both removal and stop breaks the case where you have multiple Waits for different scenarios and the result is overwritten by the other.
> 
> 😢 gotcha; my bad!

Noo! 🤣 I should have remembered about this. I actually hit this problem with my initial version of the fix but I forgot about that.

corhere commented:
> Regarding making the `Wait` synchronous - this involves much more code to be changed, so that's possibly something that could be addressed in a follow-up. However I'm not sure if this wouldn't actually encourage a racy approach to calling it. Usually `Wait` involves some kind of action that will actually make the state transition.

You make a very compelling argument. However, nearly all the callers immediately receive on the channel _in the very same line the method is called_. It is already used like a synchronous function in practice so you might be surprised how little would need to change on the caller side if `func (*State) Wait` was changed to be synchronous. From what I can see, pretty much only `func (*Daemon) ContainerWait` would need to spawn a goroutine to wait asynchronously. Both the implementation and all other callers would be simplified. I think it would be to your benefit to make the switch as part of this PR, but I'll respect your decision to defer that change to a follow-up.

In fact, I think you actually demonstrated that the `Wait(WaitConditionNextExit)` API is fundamentally broken! A `Wait`-like API works well to synchronize with a level-sensitive condition, such as `WaitConditionNotRunning`, where it does not matter whether the wait started before or after the transition to the target state. (`WaitConditionRemoved` is weird; "removed" is logically a terminal state so should be level-sensitive, but it has historically been implemented as an edge-sensitive wait. I think that was a bug which you have now mostly fixed.) As you have astutely pointed out, using such an API to wait for an edge-sensitive condition such as `WaitConditionNextExit` is inherently racy without special care because wait needs to be started before performing the action which causes the state transition. And even _with_ special care it is still racy! The state transition could still be triggered asynchronously by something else which races the waiter-to-be. For example, the container process could exit before `Wait(WaitConditionNextExit)` is called. (By the same logic, the `POST /containers/{name}/wait` API is also fundamentally broken. Improving the Engine API is _definitely_ out of scope for this PR. Its API handler also happens to be the only place where `WaitConditionNextExit` is referenced outside of tests.)

corhere reviewed (changes requested):
I think this is a great improvement over the previous iteration as it is very explicit and _not_ clever. But I'm biased.

vvoland committed:
daemon/restart: Don't mutate AutoRemove when restarting

This caused a race condition where AutoRemove could be restored before
container was considered for restart and made autoremove containers
impossible to restart.

```
$ make DOCKER_GRAPHDRIVER=vfs BIND_DIR=. TEST_FILTER='TestContainerWithAutoRemoveCanBeRestarted' TESTFLAGS='-test.count 1' test-integration
...
=== RUN   TestContainerWithAutoRemoveCanBeRestarted
=== RUN   TestContainerWithAutoRemoveCanBeRestarted/kill
=== RUN   TestContainerWithAutoRemoveCanBeRestarted/stop
--- PASS: TestContainerWithAutoRemoveCanBeRestarted (1.61s)
    --- PASS: TestContainerWithAutoRemoveCanBeRestarted/kill (0.70s)
    --- PASS: TestContainerWithAutoRemoveCanBeRestarted/stop (0.86s)
PASS

DONE 3 tests in 3.062s
```

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland commented on container/state.go:
```diff
@@ -33,8 +33,12 @@ type State struct {
 
-	waitStop   chan struct{}
-	waitRemove chan struct{}
+	waitState chan waiters
+}
+
+type waiters struct {
+	stopWaiters       []chan<- StateStatus
+	removeOnlyWaiters []chan<- StateStatus
```
Good point, done. This mutex outside visibility (as well as the whole API) could really use some redesign, but this is definitely out of scope of this PR.

vvoland commented on container/state.go:
```diff
@@ -308,8 +305,37 @@ func (s *State) SetRestarting(exitStatus *ExitStatus) {
+	s.notifyWaiters(false)
+}
+
+func (s *State) notifyWaiters(removed bool) {
```
Yup, I tried to be quick about it so just decided to go with a shortcut here, since this is only a 'private' function 😄 
Done it in a bit more explicit way.

vvoland commented on container/state.go:
```diff
@@ -308,8 +305,37 @@ func (s *State) SetRestarting(exitStatus *ExitStatus) {
+	if removed {
+		for _, c := range allWaiters.removeOnlyWaiters {
+			c <- result
+			close(c)
```
Done

vvoland commented on container/state.go:
```diff
@@ -195,20 +201,20 @@ func (s *State) Wait(ctx context.Context, condition WaitCondition) <-chan StateS
 		return resultC
 	}
```
Done

vvoland commented:
Applied suggestions from @corhere, rebased on latest master and squashed the commits.

thaJeztah commented on container/state.go:
```diff
@@ -59,10 +60,7 @@ func (s StateStatus) Err() error {
 
 // NewState creates a default state object with a fresh channel for state changes.
```
Is the "with a fresh channel for state changes." bit still correct now?

(just a quick glance over the changes - haven't fully reviewed yet)

vvoland replied:
Nope, it's outdated. Fixed, thanks!

vvoland committed:
state/Wait: Fix race when reading exit status

Before this change there was a race condition between State.Wait reading
the exit code from State and the State being changed instantly after the
change which ended the State.Wait.

Now, each State.Wait has its own channel which is used to transmit the
desired StateStatus at the time the state transitions to the awaited
one. Wait no longer reads the status by itself so there is no race.

The issue caused the `docker run --restart=always ...' to sometimes exit
with 0 exit code, because the process was already restarted by the time
State.Wait got the chance to read the exit code.

Test run
--------
Before:
```
$ go test -count 1 -run TestCorrectStateWaitResultAfterRestart .
--- FAIL: TestCorrectStateWaitResultAfterRestart (0.00s)
    state_test.go:198: expected exit code 10, got 0
FAIL
FAIL    github.com/docker/docker/container      0.011s
FAIL

```

After:
```
$ go test -count 1 -run TestCorrectStateWaitResultAfterRestart .
ok      github.com/docker/docker/container      0.011s
```

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

corhere reviewed (approved):
LGTM!

vvoland commented on integration/container/restart_test.go:
```diff
@@ -153,3 +154,55 @@ func pollForNewHealthCheck(ctx context.Context, client *client.Client, startTime
+				testContainer.WithName("autoremove-restart-and-"+tc.desc),
+				testContainer.WithAutoRemove,
+			)
+			defer cli.ContainerRemove(ctx, cID, types.ContainerRemoveOptions{Force: true})
```
Done, thank you!

vvoland commented:
@corhere: I was working on follow-up with changing the Wait to be synchronous and after fixing some tests I'm more inclined about leaving it the way it is now.

The way to use this synchronous Wait in an asynchronous context, that I previously suggested is actually incorrect 😆  
```
c := make(chan Result)
go func () {
  c <- ctr.WaitSynchronous(ctx, WaitConditionNextExit)
}()
client.ContainerRestart(ctx, containerID)
<- c
```

What's the problem here? The Wait is running in another goroutine, so the client.ContainerRestart will actually run before the WaitSynchronous has the chance to setup the wait.
So at least we'd need something like this:
```
c := make(chan Result)
go func () {
  c <- ctr.WaitSynchronous(ctx, WaitConditionNextExit)
}()
runtime.Gosched() // or some hacky time.Sleep(but how much to wait for?)
client.ContainerRestart(ctx, containerID)
<- c
```

Even then - I think it won't always be correct because the WaitSynchronous locks a mutex, so it may also need to wait. Also other running/waiting goroutines may have some impact on whether the new goroutine will be scheduled before the 'main' code continues.
So with the current API I think we need to leave the Wait asynchronous as it is.

vvoland commented:
@thaJeztah LGTY?

thaJeztah reviewed (approved):
LGTM, thanks!
